waspc/src/Wasp/Util/IO.hs

```diff
@@ -122,3 +126,22 @@ isDirectoryEmpty :: Path' Abs (Dir d) -> IO Bool
 isDirectoryEmpty dirPath = do
   (files, dirs) <- listDirectory dirPath
   return $ null files && null dirs
+
+-- | Takes number of tries so far and returns number of microseconds to pause before next try.
+type RetryDelayStrategy = (Int -> Int)
+
+-- | Runs given action, and then if it fails, retries it, up to maxNumRetries.
+--   Uses provided delayStrategy to calculate pause between tries.
+-- TODO: Write tests.
+retry :: (MonadUnliftIO m) => RetryDelayStrategy -> Int -> m (Either e a) -> m (Either e a)
+retry delayStrategy maxNumRetries action = go 0
+  where
+    go numFailedRetries = do
+      action >>= \case
+        Right result -> pure $ Right result
+        Left e ->
+          if numFailedRetries < maxNumRetries
+            then do
+              liftIO $ threadDelay $ delayStrategy (numFailedRetries + 1)
+              go (numFailedRetries + 1)
+            else pure $ Left e
```

Comment:
I keep confusing myself with `numFailedRetries` and if it should be called `numFailedTries` or not. On L139 it's the number of tries so far, on L143 it's the number of retries so far, and on L145 & 146 `numFailedRetries + 1` is again the number of tries so far.

I think the function is understandable how it is, it just took me a while to convince myself that there is no off-by-one error in the number of retries it will do.

Reply:
You are right, I rewrote it something and I think now it is cleaner, take a look pls.

Reply:
Looks good now!